### PR TITLE
Return early from draw calls on layers with constant opacity=0

### PR DESCRIPTION
**Pull request by lbud**
We're completing full draw calls for layers even when their opacity is set to zero. This cuts back on render passes by bailing early from draw functions if a layer's opacity is constant at 0. One common use case for this is switching very quickly between several visualization layers without having to reparse buckets/re-download tiles.

Locally I replaced the Streets style in the Paint benchmark with a version where _half_ of the layers are set to opacity=0 — obviously this example is extremely contrived, but
image

### CODE REVIEW

File: src/render/draw_circle.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ module.exports = drawCircles;
 
 function drawCircles(painter: Painter, sourceCache: SourceCache, layer: CircleStyleLayer, coords: Array<TileCoord>) {
     if (painter.renderPass !== 'translucent') return;
+    if (layer.paint['circle-opacity'] === 0 && layer.paint['circle-stroke-opacity'] === 0) return;
```
Comment (anandthakker) [inline src/render/draw_circle.js]: With an eye towards https://github.com/mapbox/mapbox-gl-js/issues/3044, we should use `layer.isPaintValueFeatureConstant()` and `layer.getPaintValue()` instead of directly accessing `layer.paint`

**Comment by lbud** [comment]
Good idea @anandthakker — moved these in https://github.com/mapbox/mapbox-gl-js/pull/5429/commits/09ed3de68501a68f1736cef550e567ff4653a11c.

**Comment by jfirebaugh** [comment]
> I wonder if it makes sense to fold these checks into `isHidden()` for each layer type?

I seem to recall that's what we used to do, but we had reason to stop. Can you please review the history and see why?

**Comment by lbud** [comment]
Good call @jfirebaugh — looks like this was removed in https://github.com/mapbox/mapbox-gl-js/pull/3120. It seems like the cause of that underlying bug was that we were also trying to be smart about the alpha channel in the color, which I think isn't necessary here — users trying to show/hide layers with opacity would do so with explicit `-opacity` properties. From that PR:

> The latest commit removes all opacity tests from the `StyleLayer#isHidden` method. This is unlikely to cause performance regressions in core styles because we don't often have 0 opacity layers in core styles. It's possible a 3rd party style may see a perf regression due to this change.

¯\_(ツ)_/¯

**Comment by jfirebaugh** [comment]
Yeah, that's the one I was thinking of. In addition to that, we also tell people to use opacity 0 if they don't want the layer to be visible, but do want it to be queryable with `queryRenderedFeatures`. I believe including an opacity check in `isHidden` would break that, because `isHidden` affects whether the tiles are loaded. So I think we should short circuit in drawing, but not change `isHidden`.

**Comment by lbud** [comment]
Okay, this now short-circuits only in draw functions with `isOpacityZero` methods per StyleLayer subclass. I've also added two query tests for invisible features (one `opacity=0` and one `visibility=none`) and a render test for circles with strokes but zero-opacity fills, since that's a slightly weird edge case here.

**Comment by mourner** [comment]
@lbud let's also run the bench for a style with no opacity=0 layers since this may affect performance (many additional `getPaintValue` calls)

**Comment by lbud** [comment]
@mourner with the regular Streets style there's no statistically significant change. Here are bench results from #Paint — ran it several times, always ± just a bit in either direction:

image

Full bench run:
image

@anandthakker and I discussed this yesterday as well — we could theoretically just use computed `layer.paint[<type>-opacity]`, but as he noted in https://github.com/mapbox/mapbox-gl-js/pull/5429#discussion_r143462457 we'd like to get away from doing that.

File: src/render/draw_background.js [hunk 2]
```diff
@@ -9,6 +9,8 @@ import type StyleLayer from '../style/style_layer';
 module.exports = drawBackground;
 
 function drawBackground(painter: Painter, sourceCache: SourceCache, layer: StyleLayer) {
```
Comment (jfirebaugh) [inline src/render/draw_background.js]: Change `layer: StyleLayer` to `layer: BackgroundStyleLayer`...

File: src/render/draw_raster.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ module.exports = drawRaster;
 
 function drawRaster(painter: Painter, sourceCache: SourceCache, layer: StyleLayer, coords: Array<TileCoord>) {
```
Comment (jfirebaugh) [inline src/render/draw_raster.js]: ...`layer: RasterStyleLayer`...

File: src/style/style_layer.js [hunk 4]
```diff
@@ -41,6 +41,7 @@ class StyleLayer extends Evented {
     _layoutFunctions: {[string]: boolean};
 
     +createBucket: (parameters: BucketParameters) => Bucket;
+    +isOpacityZero: (zoom: number, property?: string) => boolean;
```
Comment (jfirebaugh) [inline src/style/style_layer.js]: ...then you can remove this...

File: src/style/style_layer/symbol_style_layer.js [hunk 5]
```diff
@@ -48,6 +48,13 @@ class SymbolStyleLayer extends StyleLayer {
         return (new SymbolBucket((parameters: any)): any);
     }
 
+    isOpacityZero(zoom: number, property?: string) {
```
Comment (jfirebaugh) [inline src/style/style_layer/symbol_style_layer.js]: ...and make `property` non-optional.

Comment (lbud) [reply src/style/style_layer/symbol_style_layer.js]: 👍 thanks @jfirebaugh!
